diff --git a/src/coins.cpp b/src/coins.cpp
index 8ff652b..e1b9861 100644
--- a/src/coins.cpp
+++ b/src/coins.cpp
@@ -9,6 +9,8 @@
 
 #include <assert.h>
 
+#include <stdexcept>
+
 /**
  * calculate number of bytes for the bitmask, and its number of non-zero bytes
  * each bit in the bitmask represents the availability of one output, but the
@@ -47,6 +49,36 @@ uint256 CCoinsView::GetBestBlock() const { return uint256(); }
 bool CCoinsView::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) { return false; }
 CCoinsViewCursor *CCoinsView::Cursor() const { return 0; }
 
+void CCoinsView::FindScriptPubKey(CCoinsViewCursor& cursor, const std::set<CScript>& setscriptNeedles, std::map<uint256, CCoins>& outResults) {
+    for ( ; cursor.Valid(); cursor.Next()) {
+        uint256 hash;
+        CCoins coins;
+        if (!cursor.GetKey(hash)) {
+            throw std::runtime_error(std::string(__func__) + ": GetKey failed");
+        } else if (!cursor.GetValue(coins)) {
+            throw std::runtime_error(std::string(__func__) + ": GetValue failed");
+        }
+        bool fFoundAny = false;
+        for (CTxOut& txo : coins.vout) {
+            if (setscriptNeedles.find(txo.scriptPubKey) == setscriptNeedles.end()) {
+                txo.SetNull();
+            } else {
+                fFoundAny = true;
+            }
+        }
+        if (!fFoundAny) {
+            continue;
+        }
+        coins.Cleanup();
+        outResults.insert(std::pair<uint256, CCoins>(hash, coins));
+    }
+}
+
+void CCoinsView::FindScriptPubKey(const std::set<CScript>& setscriptNeedles, std::map<uint256, CCoins>& outResults) {
+    std::unique_ptr<CCoinsViewCursor> pcursor(Cursor());
+    FindScriptPubKey(*pcursor, setscriptNeedles, outResults);
+}
+
 
 CCoinsViewBacked::CCoinsViewBacked(CCoinsView *viewIn) : base(viewIn) { }
 bool CCoinsViewBacked::GetCoins(const uint256 &txid, CCoins &coins) const { return base->GetCoins(txid, coins); }
diff --git a/src/coins.h b/src/coins.h
index 033651a..14ed5cd 100644
--- a/src/coins.h
+++ b/src/coins.h
@@ -16,6 +16,9 @@
 #include <assert.h>
 #include <stdint.h>
 
+#include <map>
+#include <set>
+
 #include <boost/foreach.hpp>
 #include <boost/unordered_map.hpp>
 
@@ -334,6 +337,10 @@ public:
     //! Retrieve the block hash whose state this CCoinsView currently represents
     virtual uint256 GetBestBlock() const;
 
+    //! Search for a given set of pubkey scripts
+    static void FindScriptPubKey(CCoinsViewCursor& cursor, const std::set<CScript>& setscriptNeedles, std::map<uint256, CCoins>& outResults);
+    void FindScriptPubKey(const std::set<CScript>& setscriptNeedles, std::map<uint256, CCoins>& outResults);
+
     //! Do a bulk modification (multiple CCoins changes + BestBlock change).
     //! The passed mapCoins can be modified.
     virtual bool BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock);
diff --git a/src/qt/rpcconsole.cpp b/src/qt/rpcconsole.cpp
index 9d518a9..e3ec7d7 100644
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -74,6 +74,7 @@ const QStringList historyFilter = QStringList()
     << "importmulti"
     << "signmessagewithprivkey"
     << "signrawtransaction"
+    << "sweepprivkeys"
     << "walletpassphrase"
     << "walletpassphrasechange"
     << "encryptwallet";
diff --git a/src/rpc/client.cpp b/src/rpc/client.cpp
index dff21c9..92125d7 100644
--- a/src/rpc/client.cpp
+++ b/src/rpc/client.cpp
@@ -63,6 +63,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "sendmany", 1 },
     { "sendmany", 2 },
     { "sendmany", 4 },
+    { "sweepprivkeys", 0 },
     { "addmultisigaddress", 0 },
     { "addmultisigaddress", 1 },
     { "addmultisigaddress", 3 },
diff --git a/src/txdb.h b/src/txdb.h
index 5b98d27..e26f143 100644
--- a/src/txdb.h
+++ b/src/txdb.h
@@ -61,7 +61,9 @@ struct CDiskTxPos : public CDiskBlockPos
     }
 };
 
-/** CCoinsView backed by the coin database (chainstate/) */
+/** CCoinsView backed by the coin database (chainstate/)
+ * Cursor requires FlushStateToDisk for consistency.
+ */
 class CCoinsViewDB : public CCoinsView
 {
 protected:
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index cfdc508..5dc9ff0 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -864,6 +864,33 @@ TxMempoolInfo CTxMemPool::info(const uint256& hash) const
     return GetInfo(i);
 }
 
+void CTxMemPool::FindScriptPubKey(const std::set<CScript>& setscriptNeedles, std::map<uint256, CCoins>& outResults) {
+    LOCK(cs);
+    for (const CTxMemPoolEntry& entry : mapTx) {
+        const CTransaction& tx = entry.GetTx();
+        bool fFoundAny = false;
+        for (const CTxOut& txo : tx.vout) {
+            if (setscriptNeedles.find(txo.scriptPubKey) != setscriptNeedles.end()) {
+                fFoundAny = true;
+                break;
+            }
+        }
+        if (!fFoundAny) {
+            continue;
+        }
+
+        const uint256& hash = tx.GetHash();
+        CCoins coins(tx, MEMPOOL_HEIGHT);
+        for (CTxOut& txo : coins.vout) {
+            if (setscriptNeedles.find(txo.scriptPubKey) == setscriptNeedles.end()) {
+                txo.SetNull();
+            }
+        }
+        coins.Cleanup();
+        outResults.insert(std::pair<uint256, CCoins>(hash, coins));
+    }
+}
+
 CFeeRate CTxMemPool::estimateFee(int nBlocks) const
 {
     LOCK(cs);
diff --git a/src/txmempool.h b/src/txmempool.h
index bdd3d52..baffcd0 100644
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -7,6 +7,7 @@
 #define BITCOIN_TXMEMPOOL_H
 
 #include <list>
+#include <map>
 #include <memory>
 #include <set>
 
@@ -23,6 +24,7 @@
 
 class CAutoFile;
 class CBlockIndex;
+class CScript;
 
 inline double AllowFreeThreshold()
 {
@@ -626,6 +628,8 @@ public:
     TxMempoolInfo info(const uint256& hash) const;
     std::vector<TxMempoolInfo> infoAll() const;
 
+    void FindScriptPubKey(const std::set<CScript>& setscriptNeedles, std::map<uint256, CCoins>& outResults);
+
     /** Estimate fee rate needed to get into the next nBlocks
      *  If no answer can be given at nBlocks, return an estimate
      *  at the lowest number of blocks where one can be given
@@ -692,6 +696,7 @@ private:
 /** 
  * CCoinsView that brings transactions from a memorypool into view.
  * It does not check for spendings by memory pool transactions.
+ * Its Cursor also doesn't work. In general, it is broken as a CCoinsView implementation outside of a few use cases.
  */
 class CCoinsViewMemPool : public CCoinsViewBacked
 {
diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
index 39dd8e1..0a170c6 100644
--- a/src/wallet/rpcdump.cpp
+++ b/src/wallet/rpcdump.cpp
@@ -75,6 +75,8 @@ std::string DecodeDumpString(const std::string &str) {
     return ret.str();
 }
 
+void ParseWIFPrivKey(const std::string strSecret, CKey& key, CPubKey& pubkey);
+
 UniValue importprivkey(const UniValue& params, bool fHelp, CRPCRequestInfo& reqinfo)
 {
     CWallet *& pwallet = reqinfo.wallet;
@@ -120,16 +122,9 @@ UniValue importprivkey(const UniValue& params, bool fHelp, CRPCRequestInfo& reqi
     if (fRescan && fPruneMode)
         throw JSONRPCError(RPC_WALLET_ERROR, "Rescan is disabled in pruned mode");
 
-    CBitcoinSecret vchSecret;
-    bool fGood = vchSecret.SetString(strSecret);
-
-    if (!fGood) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
-
-    CKey key = vchSecret.GetKey();
-    if (!key.IsValid()) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key outside allowed range");
-
-    CPubKey pubkey = key.GetPubKey();
-    assert(key.VerifyPubKey(pubkey));
+    CKey key;
+    CPubKey pubkey;
+    ParseWIFPrivKey(strSecret, key, pubkey);
     CKeyID vchAddress = pubkey.GetID();
     {
         pwallet->MarkDirty();
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index 41875da..8484b9f 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -6,6 +6,7 @@
 #include "amount.h"
 #include "base58.h"
 #include "chain.h"
+#include "consensus/validation.h"
 #include "core_io.h"
 #include "init.h"
 #include "main.h"
@@ -55,6 +56,22 @@ void EnsureWalletIsUnlocked(CWallet * const pwallet)
         throw JSONRPCError(RPC_WALLET_UNLOCK_NEEDED, "Error: Please enter the wallet passphrase with walletpassphrase first.");
 }
 
+void ParseWIFPrivKey(const std::string strSecret, CKey& key, CPubKey& pubkey)
+{
+    CBitcoinSecret vchSecret;
+    if (!vchSecret.SetString(strSecret)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
+    }
+
+    key = vchSecret.GetKey();
+    if (!key.IsValid()) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key outside allowed range");
+    }
+
+    pubkey = key.GetPubKey();
+    assert(key.VerifyPubKey(pubkey));
+}
+
 void WalletTxToJSON(const CWalletTx& wtx, UniValue& entry)
 {
     int confirms = wtx.GetDepthInMainChain();
@@ -1001,6 +1018,154 @@ UniValue sendmany(const UniValue& params, bool fHelp, CRPCRequestInfo& reqinfo)
     return wtx.GetHash().GetHex();
 }
 
+UniValue sweepprivkeys(const UniValue& params, bool fHelp, CRPCRequestInfo& reqinfo)
+{
+    CWallet *& pwallet = reqinfo.wallet;
+
+    if (!EnsureWalletIsAvailable(pwallet, fHelp))
+        return NullUniValue;
+
+    if (fHelp || params.size() != 1)
+        throw runtime_error(
+            "sweepprivkeys {\"privkeys\": [\"bitcoinprivkey\",...], other options}\n"
+            "\nSends bitcoins controlled by private key to specified destinations.\n"
+            "\nOptions:\n"
+            "  \"privkeys\":[\"bitcoinprivkey\",...]   (array of strings, required) An array of WIF private key(s)\n"
+            "  \"label\":\"actuallabelname\"           (string, optional) Label for received bitcoins\n"
+            "  \"comment\":\"description\"             (string, optional) Local comment for the receive transaction\n"
+        );
+
+    // NOTE: It isn't safe to sweep-and-send in a single action, since this would leave the send missing from the transaction history
+
+    RPCTypeCheck(params, boost::assign::list_of(UniValue::VOBJ));
+
+    // Parse options
+    std::set<CScript> setscriptSearch;
+    CBasicKeyStore tempKeystore;
+    CMutableTransaction tx;
+    std::string strLabel, strComment;
+    CAmount nTotalIn = 0;
+    for (const std::string& optname : params[0].getKeys()) {
+        const UniValue& optval = params[0][optname];
+        if (optname == "privkeys") {
+            const UniValue& privkeys_a = optval.get_array();
+            for (size_t privkey_i = 0; privkey_i < privkeys_a.size(); ++privkey_i) {
+                const UniValue& privkey_wif = privkeys_a[privkey_i];
+                std::string strSecret = privkey_wif.get_str();
+                CKey key;
+                CPubKey pubkey;
+                ParseWIFPrivKey(strSecret, key, pubkey);
+
+                tempKeystore.AddKey(key);
+                CKeyID vchAddress = pubkey.GetID();
+                CScript script = GetScriptForDestination(vchAddress);
+                if (!script.empty()) {
+                    setscriptSearch.insert(script);
+                }
+                script = GetScriptForRawPubKey(pubkey);
+                if (!script.empty()) {
+                    setscriptSearch.insert(script);
+                }
+            }
+        } else if (optname == "label") {
+            strLabel = AccountFromValue(optval.get_str());
+        } else if (optname == "comment") {
+            strComment = optval.get_str();
+        } else {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Unrecognised option '%s'", optname));
+        }
+    }
+
+    // Ensure keypool is filled if possible
+    {
+        LOCK2(cs_main, pwallet->cs_wallet);
+
+        if (!pwallet->IsLocked()) {
+            pwallet->TopUpKeyPool();
+        }
+    }
+
+    // Reserve the key we will be using
+    CReserveKey reservekey(pwallet);
+    CPubKey vchPubKey;
+    if (!reservekey.GetReservedKey(vchPubKey)) {
+        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+    }
+
+    // Scan UTXO set for inputs
+    std::vector<CTxOut> vInTXOs;
+    {
+        // Collect all possible inputs
+        std::map<uint256, CCoins> mapcoins;
+        {
+            LOCK(cs_main);
+            mempool.FindScriptPubKey(setscriptSearch, mapcoins);
+            FlushStateToDisk();
+            pcoinsTip->FindScriptPubKey(setscriptSearch, mapcoins);
+        }
+
+        // Add them as inputs to the transaction, and count the total value
+        for (auto& it : mapcoins) {
+            const uint256& hash = it.first;
+            const CCoins& coins = it.second;
+            for (size_t txo_n = 0; txo_n < coins.vout.size(); ++txo_n) {
+                const CTxOut& txo = coins.vout[txo_n];
+                if (txo.IsNull()) {
+                    continue;
+                }
+                tx.vin.emplace_back(hash, txo_n);
+                vInTXOs.push_back(txo);
+                nTotalIn += txo.nValue;
+            }
+        }
+    }
+
+    if (nTotalIn == 0) {
+        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "No value to sweep");
+    }
+
+    CKeyID keyID = vchPubKey.GetID();
+    CTxDestination txdest = CBitcoinAddress(keyID).Get();
+
+    tx.vout.emplace_back(nTotalIn, GetScriptForDestination(txdest));
+
+    while (true) {
+        if (tx.vout[0].IsDust(::minRelayTxFee)) {
+            throw JSONRPCError(RPC_VERIFY_REJECTED, "Swept value would be dust");
+        }
+        for (size_t nIn = 0; nIn < tx.vin.size(); ++nIn) {
+            SignatureData sigdata;
+            if (!ProduceSignature(MutableTransactionSignatureCreator(&tempKeystore, &tx, nIn, vInTXOs[nIn].nValue, SIGHASH_ALL), vInTXOs[nIn].scriptPubKey, sigdata)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Failed to sign");
+            }
+            UpdateTransaction(tx, nIn, sigdata);
+        }
+        int64_t nBytes = GetVirtualTransactionSize(tx);
+        CAmount nFeeNeeded = pwallet->GetMinimumFee(nBytes, nTxConfirmTarget, mempool);
+        const CAmount nTotalOut = tx.vout[0].nValue;
+        if (nFeeNeeded <= nTotalIn - nTotalOut) {
+            break;
+        }
+        tx.vout[0].nValue = nTotalIn - nFeeNeeded;
+    }
+
+    CTransaction txFinal(tx);
+    pwallet->SetAddressBook(keyID, strLabel, "receive");
+
+    CValidationState state;
+    if (!AcceptToMemoryPool(mempool, state, txFinal, true, NULL, maxTxFee, std::set<std::string>())) {
+        pwallet->DelAddressBook(keyID);
+        if (state.IsInvalid()) {
+            throw JSONRPCError(RPC_TRANSACTION_REJECTED, strprintf("%i: %s", state.GetRejectCode(), state.GetRejectReason()));
+        } else {
+            throw JSONRPCError(RPC_TRANSACTION_ERROR, state.GetRejectReason());
+        }
+    }
+    reservekey.KeepKey();
+
+    return txFinal.GetHash().GetHex();
+}
+
 // Defined in rpc/misc.cpp
 extern CScript _createmultisig_redeemScript(const UniValue& params, CRPCRequestInfo&, bool fSorted);
 
@@ -3016,6 +3181,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "sendfrom",                 &sendfrom,                 false },
     { "wallet",             "sendmany",                 &sendmany,                 false },
     { "wallet",             "sendtoaddress",            &sendtoaddress,            false },
+    { "wallet",             "sweepprivkeys",            &sweepprivkeys,            false },
     { "wallet",             "setaccount",               &setaccount,               true  },
     { "wallet",             "settxfee",                 &settxfee,                 true  },
     { "wallet",             "signmessage",              &signmessage,              true  },
